demos/defender/defender_demo/metrics_collector.c

```diff
@@ -0,0 +1,381 @@
+    {
+        LogError( ( "Invalid parameters. pOutConnectionsArray: %p,"
+                    " connectionsArrayLength: %u, pOutNumEstablishedConnections: %p.",
+                    pOutConnectionsArray,
+                    connectionsArrayLength,
+                    pOutNumEstablishedConnections ) );
+        status = MetricsCollectorBadParameter;
+    }
+
+    if( status == MetricsCollectorSuccess )
+    {
+        fileHandle = fopen( "/proc/net/tcp", "r" );
+
+        if( fileHandle == NULL )
+        {
+            LogError( ( "Failed to open /proc/net/tcp." ) );
+            status = MetricsCollectorFileOpenFailed;
+        }
+    }
+
+    if( status == MetricsCollectorSuccess )
+    {
+        while( fgets( &( lineBuffer[ 0 ] ), MAX_LINE_LENGTH, fileHandle ) != NULL )
+        {
+            lineNumber++;
+
+            LogDebug( ( "File: /proc/net/tcp, Line: %u, Content: %s.",
+                        lineNumber,
+                        &( lineBuffer[ 0 ] ) ) );
+
+            /* Skip the first line as it is a header. */
+            if( lineNumber <= 1 )
+            {
+                continue;
+            }
+
+            /* Parse the output. */
+            filledVariables = sscanf( &( lineBuffer[ 0 ] ),
+                                      "%*[^:]: %8x:%4x %8x:%4x %2x",
+                                      &( localIp ),
+                                      &( localPort ),
+                                      &( remoteIp ),
+                                      &( remotePort ),
+                                      &( connectionStatus ) );
+
+            /* sscanf should fill all the 5 variables successfully. */
+            if( filledVariables != 5 )
+            {
+                LogError( ( "Failed to parse %s.", &( lineBuffer[ 0 ] ) ) );
+                status = MetricsCollectorParsingFailed;
+                break;
+            }
+
+            if( connectionStatus == CONNECTION_STATUS_ESTABLISHED )
+            {
+                if( pOutConnectionsArray != NULL )
+                {
+                    /* The output array member to fill. */
+                    pEstablishedConnection = &( pOutConnectionsArray[ numEstablishedConnections ] );
+
+                    pEstablishedConnection->localIp = htonl( localIp );
+                    pEstablishedConnection->remoteIp = htonl( remoteIp );
+                    pEstablishedConnection->localPort = ( uint16_t ) localPort;
+                    pEstablishedConnection->remotePort = ( uint16_t ) remotePort;
+
+                    numEstablishedConnections++;
+
+                    /* Break if the output array is full. */
+                    if( connectionsArrayLength == numEstablishedConnections )
+                    {
+                        break;
+                    }
+                }
+                else
+                {
+                    numEstablishedConnections++;
+                }
+            }
+        }
+    }
```

Comment:
If there was some problem reading the file in `fgets()`, `status` is still set to `MetricsCollectorSuccess`.  Is that correct?  Same comment in other loops which read from `/proc/net`.

Reply:
If the file open with read mode is successful, the only error fgets can get is no data - this is okay as this meas no open ports/established connections.